File: content/tracks/javascript/concepts/01-variables-and-types/variables_and_types_test.js
const assert = require("assert");
const { typeLabel, strictEquals, clamp } = require("./stub");

let passed = 0, failed = 0;
function test(name, fn) {
  try { fn(); console.log(`  PASS: ${name}`); passed++; }
  catch (e) { console.log(`  FAIL: ${name} — ${e.message}`); failed++; }
}

test("typeLabel: string",    () => assert.strictEqual(typeLabel("hi"),        "string"));
test("typeLabel: number",    () => assert.strictEqual(typeLabel(42),          "number"));
test("typeLabel: boolean",   () => assert.strictEqual(typeLabel(true),        "boolean"));
test("typeLabel: undefined", () => assert.strictEqual(typeLabel(undefined),   "undefined"));

test("strictEquals: 1 === 1",      () => assert.strictEqual(strictEquals(1, 1),     true));
test("strictEquals: 1 === '1'",    () => assert.strictEqual(strictEquals(1, "1"),   false));
test("strictEquals: null === null", () => assert.strictEqual(strictEquals(null, null), true));
test("strictEquals: 0 === false",  () => assert.strictEqual(strictEquals(0, false), false));

test("clamp: below min",   () => assert.strictEqual(clamp(-5, 0, 10), 0));
test("clamp: above max",   () => assert.strictEqual(clamp(15, 0, 10), 10));
test("clamp: within range", () => assert.strictEqual(clamp(5,  0, 10), 5));
test("clamp: at min",      () => assert.strictEqual(clamp(0,  0, 10), 0));
test("clamp: at max",      () => assert.strictEqual(clamp(10, 0, 10), 10));

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/python/concepts/01-variables-and-types/variables_and_types_test.py
import sys
from stub import type_label, clamp, first_number

passed = 0
failed = 0


def assert_eq(actual, expected):
    assert actual == expected, f"expected {expected!r}, got {actual!r}"


def test(name: str, fn):
    global passed, failed
    try:
        fn()
        print(f"  PASS: {name}")
        passed += 1
    except Exception as e:
        print(f"  FAIL: {name} — {e}")
        failed += 1


# type_label
test("type_label: int",      lambda: assert_eq(type_label(42), "int"))
test("type_label: str",      lambda: assert_eq(type_label("hi"), "str"))
test("type_label: float",    lambda: assert_eq(type_label(3.14), "float"))
test("type_label: bool",     lambda: assert_eq(type_label(True), "bool"))
test("type_label: NoneType", lambda: assert_eq(type_label(None), "NoneType"))
test("type_label: list",     lambda: assert_eq(type_label([]), "list"))

# clamp
test("clamp: below min",    lambda: assert_eq(clamp(-5, 0, 10), 0))
test("clamp: above max",    lambda: assert_eq(clamp(15, 0, 10), 10))
test("clamp: within range", lambda: assert_eq(clamp(5, 0, 10), 5))
test("clamp: at min",       lambda: assert_eq(clamp(0, 0, 10), 0))
test("clamp: at max",       lambda: assert_eq(clamp(10, 0, 10), 10))
test("clamp: floats",       lambda: assert_eq(clamp(1.5, 1.0, 2.0), 1.5))

# first_number
test("first_number: found",          lambda: assert_eq(first_number("foo 42 bar"), 42))
test("first_number: negative",       lambda: assert_eq(first_number("temp -7 celsius"), -7))
test("first_number: none",           lambda: assert_eq(first_number("no digits here"), None))
test("first_number: leading number", lambda: assert_eq(first_number("99 bottles"), 99))
test("first_number: empty string",   lambda: assert_eq(first_number(""), None))

print(f"\n{passed} passed, {failed} failed")
if failed:
    sys.exit(1)
